core/mm/file.c

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, Linaro Limited
+ */
+
+#include <kernel/panic.h>
+#include <kernel/refcount.h>
+#include <mm/file.h>
+#include <mm/fobj.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <types_ext.h>
+#include <util.h>
+
+/*
+ * struct file - file resources
+ * @tag:	Tag or hash uniqely identifying a the file
+ * @refc:	Reference counter
+ * @link:	Linked list element
+ * @num_slices:	Number of elements in the @slices array below
+ * @slices:	Array of file slices holding the fobjs of this file
+ *
+ * A file is constructed of slices which may be shared in different
+ * mappings/contexts. There may be holes in the file for ranges of the file
+ * that can't be shared.
+ */
+struct file {
+	uint8_t tag[FILE_TAG_SIZE];
+	struct refcount refc;
+	TAILQ_ENTRY(file) link;
+	unsigned int num_slices;
+	struct file_slice slices[];
+};
+
+static struct mutex file_mu = MUTEX_INITIALIZER;
+static TAILQ_HEAD(, file) file_head = TAILQ_HEAD_INITIALIZER(file_head);
+
+static int file_tag_cmp(const struct file *f, const uint8_t *tag,
+			unsigned int taglen)
+{
+	size_t n = 0;
+	int rc = memcmp(tag, f->tag, taglen);
+
+	for (n = taglen; !rc && n < sizeof(f->tag); n++)
```

Comment:
This expects recorded tags do not end with zeros. I guess this layer does not want to put constraint on tags. Maybe store also to tag length.

Reply:
Fair enough, the code will also be easier to read if I'm using an explicit taglen instead of just checking that the zero padding is where it's supposed to be. I'll fix.